### ISSUE TITLE
test_coax_two_port_ad fd_consistent is red on main: rel=0.02631 vs a 0.02 threshold

### ISSUE THREAD

**Issue by bk-squared** [bk-squared/rfx#665]
`tests/test_coax_two_port_ad.py -k fd_consistent` fails on `main`:

```
assert 0.02630975141316488 <= 0.02
```

Reproduced on a pinned `git archive` of `fd37c62` in a clean scratch tree — **to the digit**, so
it is deterministic here and not a flake or a contaminated worktree. Found while verifying an
unrelated PR's claim that the red was pre-existing; it was.

The overshoot is 31% above the threshold, which is more than a near-margin platform wobble but
small enough that it is plausibly an envelope question rather than a broken gradient. Two things
worth separating before anyone widens anything:

- Is the AD gradient wrong, or is the FD reference the noisy side? This repo has a recorded case
  (#527) where an apparently-failing AD gate turned out to be a float32 comparator, and the AD
  side was vindicated once the comparison arithmetic moved to float64.
- Is 0.02 a measured envelope or a round number? If it was set from a measurement, on what
  platform and at what precision.

**Do not widen the threshold to make this green without a written root cause** — this repo's
standing rule, and the #527 precedent is exactly why.

Related but not the same: #610 covers near-margin bounds in `test_msl_port_integration`.

**Comment by bk-squared** [comment]
Reproduction data from the 2026-08-23 audit (finding E-4): at this issue's own pinned commit fd37c62 AND at current HEAD, the test PASSES standalone on the audit machine — rel_err 0.00599 against the 0.02 threshold, bit-identical intermediates between the two commits. Separately, the same test FAILED today inside a 3-way pytest-split shard (group 1 of `-m slow_physics`) on an archive of the same tree. Three observations — 0.006 standalone, a shard failure, and this issue's original 0.02631 — are consistent with ORDER/STATE dependence (accumulated XLA or x64 state inside a long shard), not with a fixed physics regression. Suggest the next step here is running it first-in-shard vs last-in-shard rather than re-deriving the gradient math.

**Comment by bk-squared** [comment]
Correction to my previous comment — the order-dependence hypothesis is WRONG, measured within the hour. Completing the matrix on one host, one venv, `git archive` scratch trees with `PYTHONPATH` pinned and `rfx.__file__` asserted:

| execution context | fix/sheet-node-permittivity | main |
|---|---|---|
| shell A, in-shard | FAIL | — |
| shell A, standalone | FAIL | FAIL |
| shell B (agent sandbox), standalone | — | PASS, rel_err 0.00599 (and PASS at fd37c62, bit-identical intermediates) |

Two facts fall out. (1) **The branch is exonerated**: branch and main behave identically in the same context, consistent with the V173-A harness where the branch is bit-inert. (2) The discriminating variable is the **execution context**, not shard order and not the commit: the same test on the same code swings rel_err 0.006 ↔ >0.02 between two shell contexts on one machine (candidate mechanisms: thread count / BLAS pool / XLA autotuning differences between an interactive shell and an agent sandbox — not yet isolated). So this gate is environment-sensitive near its 0.02 threshold, and this issue's 'red on main' is better read as 'red in some environments'. Next probe that would settle it: run once with `OPENBLAS_NUM_THREADS=1 XLA_FLAGS=--xla_cpu_multi_thread_eigen=false` in the failing context and see which side it lands on.

**Comment by bk-squared** [comment]
Probe result (the next-probe from my earlier comment, run): thread pinning does NOT flip it. Same failing shell context, same main archive, standalone: unpinned FAIL, `OPENBLAS_NUM_THREADS=1 XLA_FLAGS=--xla_cpu_multi_thread_eigen=false` FAIL. So the context discriminator is not shard order (earlier correction), not the commit, and now not the thread count. Ruled out so far: branch, shard order, threads. Still open: what else differs between an interactive shell and the agent-sandbox context where rel_err = 0.00599 passes — candidate next probe: dump `os.environ` + `jax.devices()`/backend flags from both contexts and diff, then bisect the env delta. Runtimes also swing with host load (56 s idle vs ~7 min loaded) which is worth watching as a confounder for the FD step quality.

**Comment by bk-squared** [comment]
Context-matrix update + a mechanism lead (2026-08-26, agent shell, jax 0.10.2 CPU, x64 off, no JAX/XLA env vars set):

**This context FAILS with the issue's exact digits**: `rel_err = 0.02631` — so the earlier shell-A/shell-B split is not "interactive fails / sandbox passes". Combined with the 0.00599 passes recorded on 2026-08-24, the picture is now **two deterministic attractors** (0.02631 and 0.00599), each exactly reproducible within its context — which smells like a discrete configuration difference (dtype/promotion path), not noise.

**New lead, measured**: running with `-W error::FutureWarning` turns a silent demotion into a traceback that lands in the forward solve's CPML update:

```
rfx/api/_sparams.py:5930 compute_coaxial_two_port
rfx/simulation.py:1282 core_step
rfx/boundaries/cpml.py:997 apply_cpml_h
FutureWarning: scatter inputs have incompatible types: cannot safely cast value from dtype=float64 to dtype=float32
```

i.e. some CPML H-update operand is float64 and gets silently truncated into the float32 field scatter — on the hot path of this exact test (#527 class: a precision seam inside the comparison chain). Candidate (UNVERIFIED — needs instrumentation, not asserting): `apply_cpml_h`'s no-materials branch sets `ch_* = dt / MU_0` where `grid.dt` is a `np.float64` scalar — strong-typed in JAX promotion, so `ch * correction` promotes f32→f64 before the demoting scatter. The materials branch measured clean f32 on a generic fixture, so the coax fixture's specific path needs the dtype dump.

Next probe (sharp, cheap): instrument `cpml.py:997`'s operand dtypes inside this test's fixture, fix the one f64 source following the #646 promote-never-pin family rule, and re-run `fd_consistent` in this failing context — if rel collapses to 0.00599 the two-attractor mystery closes with a root cause and the threshold never needs touching.

**Comment by bk-squared** [comment]
Root-caused and fixed by the merged PR: grid.dt (np.float64, strong-typed in JAX promotion) promoted the CPML ch coefficients to f64 on jnp-material paths; the per-step field scatter demoted them back (the FutureWarning). Weak-typing dt makes the coefficients follow the ambient dtype. Measured in this issue's red context: rel_err 0.02631 -> 0.00860 PASS, warning gone; threshold untouched. Note: the old passing attractor was 0.00599 — the fix changes the arithmetic (f32 throughout vs f64-then-demoted per step), so that value's archaeology would need the original context's JAX version; the seam itself is measured, fixed, and regression-covered (CPML oracle/asymmetric/pad/PMC 16/16). Merge was deliberately sequenced AFTER the running crossval pattern job checked out its code, keeping that comparison chain on one arithmetic.
